File: LeetCode/409. Longest Palindrome.cpp
// https://leetcode.com/problems/longest-palindrome/
class Solution
{
public:
    int longestPalindrome(string s)
    {
        unordered_map<char, int> hashMap;
        for (auto &c : s)
        {
            if (hashMap.find(c) == hashMap.end())
            {
                hashMap[c] = 1;
            }
            hashMap[c] += 1;
        }

        int cnt = 0;
        for (auto &e : hashMap)
        {
            if (e.second % 2 == 0)
            {
                ++cnt;
            }
        }
        if (cnt > 0)
        {
            cnt -= 1;
        }

        return s.length() - cnt;
    }
};

File: LeetCode/Python3/409.LongestPalindrome.py
# https://leetcode.com/problems/longest-palindrome/

# Time Complexity O(n)
# Space Complexity O(128)
# 先统计，然后成对计算
class Solution:
    def longestPalindrome(self, s: str) -> int:
        freqs = [0]*128
        
        for c in s:
            freqs[ord(c)]+=1 
            
        ans , odd = 0, 0
        for freq in freqs:
            ans += freq & (~1)
            odd |= freq & (1)
            
        return ans + odd
            
